precompiles/staking/staking.go

```diff
@@ -0,0 +1,429 @@
+package staking
+
+import (
+	"fmt"
+	"math/big"
+
+	"cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingkeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+)
+
+const (
+	// write
+	StakeMethodName     = "stake"
+	UnstakeMethodName   = "unstake"
+	MoveStakeMethodName = "moveStake"
+
+	// read
+	GetAllValidatorsMethodName = "getAllValidators"
+	GetSharesMethodName        = "getShares"
+)
+
+var (
+	ABI                 abi.ABI
+	ContractAddress     = common.HexToAddress("0x0000000000000000000000000000000000000066")
+	GasRequiredByMethod = map[[4]byte]uint64{}
+	ViewMethod          = map[[4]byte]bool{}
+)
+
+func init() {
+	initABI()
+}
+
+func initABI() {
+	if err := ABI.UnmarshalJSON([]byte(IStakingMetaData.ABI)); err != nil {
+		panic(err)
+	}
+
+	GasRequiredByMethod = map[[4]byte]uint64{}
+	for methodName := range ABI.Methods {
+		var methodID [4]byte
+		copy(methodID[:], ABI.Methods[methodName].ID[:4])
```

Comment:
What's inside `ABI.Methods[methodName].ID`? Is it just a raw representation of the method's signature, or has it already been hashed with keccak256?

Also, let's unify this as `type funcSelector ...` with some helpers to avoid code duplication.

Reply:
> its what we get from go bindings, first 4 bytes is function signature

Eth says its should be `keccak256(signature)[:4]` not `signature[:4]`. Can you please ensure that?

https://docs.soliditylang.org/en/develop/abi-spec.html#function-selector